ruby/lib/google/protobuf/repeated_field.rb

```diff
@@ -29,11 +29,127 @@
+        :pack, :permutation, :product, :pretty_print, :pretty_print_cycle,
+        :rassoc, :repeated_combination, :repeated_permutation, :reverse, :rindex, :rotate,
+        :sample, :shuffle, :shelljoin, :slice, :to_s, :transpose, :uniq, :|
+
+
+      def first(n=nil)
+        n ? self[0..n] : self[0]
+      end
+
+
+      def last(n=nil)
+        n ? self[(self.size-n-1)..-1] : self[-1]
+      end
+
+
+      def pop(n=nil)
+        if n
+          results = []
+          n.times{ results << pop_internal }
+          return results
+        else
+          return pop_internal
+        end
+      end
+
+
+      def empty?
+        self.size == 0
+      end
+
+      # array aliases into enumerable
+      alias_method :each_index, :each_with_index
+      alias_method :slice, :[]
+      alias_method :values_at, :select
+      alias_method :map, :collect
+
+
+      %w(delete delete_at delete_if shift slice! unshift).each do |method_name|
+        define_method(method_name) do |*args, &block|
+          arr = self.to_a
+          result = arr.send(method_name, *args)
+          self.replace(arr)
+          return result if result
+          return block ? block.call : result
+        end
+      end
+
+
+      %w(collect! compact! fill flatten! insert reverse! rotate! select! shuffle! sort! sort_by! uniq!).each do |method_name|
+        define_method(method_name) do |*args, &block|
+          # result can be an Enumerator, Array, or nil
+          # Enumerator can sometimes be returned if a block is an optional argument and it is not passed in
+          # nil usually specifies that no change was made
+          result = self.to_a.send(method_name, *args, &block)
+          if result
+            new_arr = result.to_a
+            self.replace(new_arr)
+            if result.is_a?(Enumerator)
+              # generate a fresh enum; rewinding the exiting one, in Ruby 2.2, will
+              # reset the enum with the same length, but all the #next calls will
+              # return nil
+              result = new_arr.to_enum
+              # generate a wrapper enum so any changes which occur by a chained
+              # enum can be captured
+              ie = InternalEnumerator.new(self, result)
+              result = ie.to_enum
+            end
+          end
+          result
+        end
+      end
+      alias_method :keep_if, :select!
+      alias_method :map!, :collect!
+      alias_method :reject!, :delete_if
+
+
+      class InternalEnumerator < Struct.new(:repeated_field, :external_enumerator)
+        def each(*args, &block)
+          orig_vals = []
+          results = []
+          external_enumerator.each do |val|
+            result = yield(val)
+            #nil means no change occured from yield; usually occurs when #to_a is called
+            result ||= val
+            orig_vals << val
+            results << result
+          end
+          repeated_field.replace(results) if results != orig_vals
```

Comment:
Wow, that's weird. It looks like it stopped in the middle of a quicksort or something -- the output is pemuted but not in any intelligible way.

It seems this actually gives us more freedom if we just assume "results are undefined if the block stops the iteration"; i.e., we can mutate self however is most convenient. Unless I'm missing some other constraint.